src/engine/Union.cpp

```diff
@@ -347,172 +346,250 @@ struct Wrapper {
-                                                       permutation);
-        co_yield pair;
+  // _____________________________________________________________________________
+  // Always inline makes makes a huge difference on large datasets.
+  AD_ALWAYS_INLINE bool isSmaller(const auto& row1, const auto& row2) const {
+    using StaticRange = std::span<const std::array<size_t, 2>, SPAN_SIZE>;
+    for (auto [index1, index2] : StaticRange{targetOrder_}) {
+      if (index1 == Union::NO_COLUMN) {
+        return true;
+      }
+      if (index2 == Union::NO_COLUMN) {
+        return false;
+      }
+      if (row1[index1] != row2[index2]) {
+        return row1[index1] < row2[index2];
       }
-    } else {
-      resultTable.insertAtEnd(it->idTable_, index, std::nullopt, permutation,
-                              Id::makeUndefined());
     }
+    return false;
+  }
+
+  // _____________________________________________________________________________
+  std::optional<Result::IdTableVocabPair> passNext(
+      const std::vector<ColumnIndex>& permutation, auto& it, auto end,
+      size_t& index) {
+    if (it != end) {
+      Result::IdTableVocabPair pair = moveOrCopy(*it);
+      pair.idTable_ = applyPermutation_(std::move(pair.idTable_), permutation);
+
+      index = 0;
+      ++it;
+      return pair;
+    }
+    return std::nullopt;
+  }
+
+  // ___________________________________________________________________________
+  void appendCurrent(const std::vector<ColumnIndex>& permutation, auto& it,
+                     size_t& index) {
+    resultTable_.insertAtEnd(it->idTable_, index, std::nullopt, permutation,
+                             Id::makeUndefined());
+    localVocab_.mergeWith(std::span{&it->localVocab_, 1});
     index = 0;
     ++it;
   }
-}
 
-// _____________________________________________________________________________
-template <int COMPARATOR_WIDTH>
-Result::Generator Union::computeResultKeepOrderImpl(
-    bool requestLaziness, auto range1, auto range2,
-    std::pair<std::shared_ptr<const Result>, std::shared_ptr<const Result>>)
-    const {
-  constexpr size_t extent =
-      COMPARATOR_WIDTH == 0 ? std::dynamic_extent : COMPARATOR_WIDTH;
-  IdTable resultTable{getResultWidth(), allocator()};
-  if (requestLaziness) {
-    resultTable.reserve(chunkSize);
-  }
-  LocalVocab localVocab;
-  auto it1 = range1.begin();
-  auto it2 = range2.begin();
-  size_t index1 = 0;
-  size_t index2 = 0;
-  auto pushRow = [this, &resultTable](bool left, const auto& row) {
-    resultTable.emplace_back();
-    for (size_t column = 0; column < resultTable.numColumns(); column++) {
-      ColumnIndex origin = _columnOrigins.at(column).at(!left);
-      resultTable.at(resultTable.size() - 1, column) =
-          origin == NO_COLUMN ? Id::makeUndefined() : row[origin];
+  // ___________________________________________________________________________
+  void appendRemaining(const std::vector<ColumnIndex>& permutation, auto& it,
```

Comment:
Does that break the laziness if there are many chunks left at the end, as it materializes a large table?

Reply:
No, this is only called when `requestLaziness` is false anyways